### ISSUE TITLE
Sign Extension in SBPFv2

### ISSUE THREAD

**Issue by Lichtso** [solana-labs/solana#32924]
#### Problem
Currently in the RBPF VM some ALU32 instructions do sign extend their 32 bit results (outputs) to 64 bit.

These operations sign extend their results from 32 bit to 64 bit:
`ADD32, SUB32, NEG32, MUL32, LMUL32, SDIV32, SREM32`

These operations do not:
`DIV32, MOD32, XOR32, OR32, AND32, LSH32, RSH32, ARSH32, MOV32, UDIV32, UREM32`

It is likely that most programs do not ever use the upper 32 bit of a `u32` or `i32` and thus this sign extension work is wasted and could be omitted.

#### Proposed Solution
Add an explicit sign extension instruction and remove implicit sign extension of results, see:
https://github.com/solana-foundation/solana-improvement-documents/pull/87/files

**Comment by dmakarov** [comment]
@nvjle, @jcivlin please, comment on this proposal.

**Comment by jcivlin** [comment]
> It is likely that most programs do not ever use the upper 32 bit of a `u32` or `i32` and thus this sign extension work is wasted and could be omitted.

We cannot make a decision on the ground of "likeliness". Either they are ever generated and then must be supported, or never generated and then must NOT be supported because we will not be able to test. 
If it is a discussion about the cost of implementing over not implementing (and somehow imitating these sign extensions) then it will be instrumental to have some metics and evaluation, and they may be different, for example, performance, or required effort, or anything reliable. And if such metrics cannot be provided now, then I'd pick either hand - and let one who will be doing this to decide - and in parallel created an issue to reevaluate the decision once we have anything that provides more info.

**Comment by nvjle** [comment]
> > It is likely that most programs do not ever use the upper 32 bit of a `u32` or `i32` and thus this sign extension work is wasted and could be omitted.
> 
> We cannot make a decision on the ground of "likeliness". Either they are ever generated and then must be supported, or never generated and then must NOT be supported because we will not be able to test. If it is a discussion about the cost of implementing over not implementing (and somehow imitating these sign extensions) then it will be instrumental to have some metics and evaluation, and they may be different, for example, performance, or required effort, or anything reliable. And if such metrics cannot be provided now, then I'd pick either hand - and let one who will be doing this to decide - and in parallel created an issue to reevaluate the decision once we have anything that provides more info.

I am in agreement with Jan here. ISA extensions and changes should be driven by a quantitative approach and where clear need is demonstrated by data, and a cost-benefit analysis (e.g., as is standard in ISA/microarchitecture teams/implementations). Gut feelings and likelihood are not reliable indicators. Overall, given our current state (new runtime work needed, new ABI work needed, Move compiler implementation, and so forth), ISA changes are not a priority. The same goes for all of the recent "proposals" of ISA changes.

**Comment by dmakarov** [comment]
> Add an explicit sign extension instruction and remove implicit sign extension of results.

This would change the semantics of ADD32, SUB32, NEG32, MUL32, LMUL32, SDIV32, SREM32 instructions.  In addition the compiler would have to generate the explicit sing extension instruction every time it emits one of the ALU32 instructions to preserve the semantics of the operations.  What would we gain from this change?

**Comment by Lichtso** [comment]
>  compiler would have to generate the explicit sing extension instruction every time it emits one of the ALU32 instructions to preserve the semantics of the operations

Not following every arithmetic instruction. Only when one casts from `i32` to `i64`. Otherwise the upper 32 bit of `i32` and `u32` are undefined behavior from the source langues perspective.
